File: python/journal/seal.py
"""Forward Secure Sealing support for journal writers.

Implements file-format sealing with deterministic synthetic keys,
matching systemd v260.1 HMAC byte ranges and tag object layout.
"""

import struct
import hmac
import hashlib
from .fss import gen_mk, gen_state0, get_epoch, evolve, RECOMMENDED_SECPAR, RECOMMENDED_SEEDLEN

TAG_LENGTH = 256 // 8
OBJECT_TYPE_TAG = 7
COMPATIBLE_SEALED = 1 << 0
COMPATIBLE_SEALED_CONTINUOUS = 1 << 2


class SealOptions:
    """Configures Forward Secure Sealing."""

    def __init__(self, seed, interval_usec, start_usec):
        if len(seed) != RECOMMENDED_SEEDLEN:
            raise ValueError(f'seal seed must be {RECOMMENDED_SEEDLEN} bytes')
        self.seed = bytes(seed)
        self.interval_usec = interval_usec
        self.start_usec = start_usec


class SealState:
    """Per-writer FSS+HMAC state."""

    def __init__(self, opts):
        self._msk, self._mpk = gen_mk(opts.seed, RECOMMENDED_SECPAR)
        self._fsprg_state = gen_state0(self._mpk, opts.seed)
        self._seed = opts.seed
        self._interval = opts.interval_usec
        self._start = opts.start_usec
        self._hmac = None
        self._hmac_running = False

    def get_epoch(self):
        return get_epoch(self._fsprg_state)

    def get_goal_epoch(self, realtime):
        if self._start == 0 or self._interval == 0:
            raise ValueError('FSS start or interval not set')
        if realtime < self._start:
            raise ValueError('realtime before FSS start')
        return (realtime - self._start) // self._interval

    def need_evolve(self, realtime):
        goal = self.get_goal_epoch(realtime)
        epoch = self.get_epoch()
        if epoch > goal:
            raise ValueError(f'FSS epoch {epoch} > goal {goal}')
        return epoch != goal

    def evolve_state(self):
        self._fsprg_state = evolve(self._fsprg_state)

    def hmac_start(self):
        if self._hmac_running:
            return
        from .fss import get_key
        key = get_key(self._fsprg_state, TAG_LENGTH, 0)
        self._hmac = hmac.new(key, digestmod=hashlib.sha256)
        self._hmac_running = True

    def hmac_write(self, data):
        self.hmac_start()
        self._hmac.update(data)

    def hmac_reset(self):
        self._hmac_running = False
        self._hmac = None

    def hmac_sum(self):
        return self._hmac.digest()

File: node/src/lib/seal.js
// Forward Secure Sealing support for journal writers.

import { createHmac } from 'node:crypto';
import {
  fsprgGenMK, fsprgGenState0, fsprgGetEpoch, fsprgEvolve, fsprgGetKey,
  RECOMMENDED_SECPAR, RECOMMENDED_SEEDLEN,
} from './fss.js';

export const TAG_LENGTH = 256 / 8;
export const OBJECT_TYPE_TAG = 7;
export const COMPATIBLE_SEALED = 1 << 0;
export const COMPATIBLE_SEALED_CONTINUOUS = 1 << 2;

export class SealOptions {
  constructor(seed, intervalUsec, startUsec) {
    if (seed.length !== RECOMMENDED_SEEDLEN) {
      throw new Error(`seal seed must be ${RECOMMENDED_SEEDLEN} bytes`);
    }
    this.seed = Buffer.from(seed);
    this.intervalUsec = intervalUsec;
    this.startUsec = startUsec;
  }
}

export class SealState {
  constructor(opts) {
    const { msk, mpk } = fsprgGenMK(opts.seed, RECOMMENDED_SECPAR);
    this.fsprgState = fsprgGenState0(mpk, opts.seed);
    this.msk = msk;
    this.seed = Buffer.from(opts.seed);
    this.interval = BigInt(opts.intervalUsec);
    this.start = BigInt(opts.startUsec);
    this.hmac = null;
    this.hmacRunning = false;
  }

  getEpoch() {
    return fsprgGetEpoch(this.fsprgState);
  }

  getGoalEpoch(realtime) {
    if (this.start === 0n || this.interval === 0n) {
      throw new Error('FSS start or interval not set');
    }
    const rt = BigInt(realtime);
    if (rt < this.start) {
      throw new Error('realtime before FSS start');
    }
    return (rt - this.start) / this.interval;
  }

  needEvolve(realtime) {
    const goal = this.getGoalEpoch(realtime);
    const epoch = this.getEpoch();
    if (epoch > goal) {
      throw new Error(`FSS epoch ${epoch} > goal ${goal}`);
    }
    return epoch !== goal;
  }

  evolveState() {
    this.fsprgState = fsprgEvolve(this.fsprgState);
  }

  hmacStart() {
    if (this.hmacRunning) return;
    const key = fsprgGetKey(this.fsprgState, TAG_LENGTH, 0);
    this.hmac = createHmac('sha256', key);
    this.hmacRunning = true;
  }

  hmacWrite(data) {
    this.hmacStart();
    this.hmac.update(data);
  }

  hmacReset() {
    this.hmacRunning = false;
    this.hmac = null;
  }

  hmacSum() {
    return this.hmac.digest();
  }
}
